File: lib-database/src/main/java/com/ustadmobile/core/db/dao/StateDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmRepository;
import com.ustadmobile.lib.db.entities.StateEntity;
import com.ustadmobile.lib.db.sync.dao.SyncableDao;

import java.util.List;

@UmDao
@UmRepository
public abstract class StateDao implements SyncableDao<StateEntity, StateDao> {

    @UmQuery("SELECT * FROM StateEntity WHERE stateId = :id AND agentUid = :agentUid AND activityId = :activityId " +
            "AND registration = :registration AND isactive LIMIT 1")
    public abstract StateEntity findByStateId(String id, long agentUid, String activityId, String registration);

    @UmQuery("SELECT * FROM StateEntity WHERE agentUid = :agentUid AND activityId = :activityId " +
            "AND registration = :registration AND isactive AND timestamp > :since")
    public abstract List<StateEntity> findStateIdByAgentAndActivity(long agentUid, String activityId, String registration, String since);

    @UmQuery("UPDATE StateEntity SET isactive = :isActive WHERE agentUid = :agentUid AND activityId = :activityId " +
            "AND registration = :registration AND isactive")
    public abstract void updateStateToInActive(long agentUid, String activityId, String registration, boolean isActive);

    @UmQuery("UPDATE StateEntity SET isactive = :isActive WHERE stateId = :stateId AND agentUid = :agentUid AND activityId = :activityId " +
            "AND registration = :registration AND isactive")
    public abstract void setStateInActive(String stateId, long agentUid, String activityId, String registration, boolean isActive);
}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/StateDao.kt
package com.ustadmobile.core.db.dao

import com.ustadmobile.door.annotation.DoorDao
import androidx.room.Query
import com.ustadmobile.door.annotation.*
import com.ustadmobile.lib.db.entities.StateEntity

@DoorDao
@Repository
expect abstract class StateDao : BaseDao<StateEntity> {

    @Query("SELECT * FROM StateEntity WHERE stateId = :id AND agentUid = :agentUid AND activityId = :activityId AND registration = :registration AND isIsactive LIMIT 1")
    abstract fun findByStateId(id: String?, agentUid: Long, activityId: String?, registration: String?): StateEntity?

    @Query("SELECT * FROM StateEntity WHERE agentUid = :agentUid AND activityId = :activityId " + "AND registration = :registration AND isIsactive AND timestamp > :since")
    @SqliteOnly
    abstract fun findStateIdByAgentAndActivity(agentUid: Long, activityId: String, registration: String, since: String): List<StateEntity>

    @Query("""
        UPDATE StateEntity 
           SET isIsactive = :isActive,
               stateLct = :updateTime
         WHERE agentUid = :agentUid AND activityId = :activityId 
           AND registration = :registration AND isIsactive
    """)
    abstract fun updateStateToInActive(
        agentUid: Long,
        activityId: String,
        registration: String,
        isActive: Boolean,
        updateTime: Long,
    )

    @Query("""
        UPDATE StateEntity 
          SET isIsactive = :isActive, 
              stateLct = :updateTime
        WHERE stateId = :stateId AND agentUid = :agentUid 
          AND activityId = :activityId AND registration = :registration 
          AND isIsactive""")
    abstract fun setStateInActive(
        stateId: String,
        agentUid: Long,
        activityId: String,
        registration: String,
        isActive: Boolean,
        updateTime: Long,
    )
}
